### PR TITLE
Run the language server in the browser

### PR DESCRIPTION
**Pull request by turbolent**
Work towards onflow/flow-playground#25

- Refactor the language server: Move the Flow/Emulator-specific features out of the language server into a separate package. At the same time, make the language server easily extensible (additional diagnostics, additional code lenses, how to resolve imports, etc.)
- Add a new entry point for the language server, specifically for running it compiled to WebAssembly in a JS environment, like the browser
- Add an example client, which loads the language server compiled to WebAssembly, and hooks it up to [Monaco](https://microsoft.github.io/monaco-editor/) using [monaco-languageclient](https://github.com/TypeFox/monaco-languageclient)

Screen Shot 2020-07-16 at 7 23 45 PM

Screen Shot 2020-07-17 at 4 29 46 PM

### CODE REVIEW

File: languageserver/client/src/language-server.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  // The callback that the language server calls
+  // to get the code for an imported address, if any
+  getAddressCode(string): string
+
+  toServer(error: any, message: any): void
+}
+
+export async function startCadenceLanguageServer(callbacks: Callbacks) {
+  const wasm = await fetch("./languageserver.wasm")
```
Comment (10thfloor) [inline languageserver/client/src/language-server.ts:24]: 💯

File: languageserver/client/src/cadence.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+      "||",
+      "?",
+      "??",
+      ":",
+      "=",
+      "@"
+    ],
+
+    // we include these common regular expressions
+    symbols: /[=><!~?:&|+\-*\/\^%]+/,
```
Comment (benjaminkvm) [inline languageserver/client/src/cadence.ts:108]: Might want to add `#` now for pragmas!

Comment (turbolent) [reply languageserver/client/src/cadence.ts:108]: Yeah, this was just copied over from the Playground and we needs to be updated

**Comment by turbolent** [comment]
I extended this a bit more based on my experience when I integrated it in the Playground: I added support for multiple language servers, and support for shutting a language server down. Would love another review 🙏

File: languageserver/client/tsconfig.json [hunk 3]
```diff
@@ -1,6 +1,9 @@
 {
     "compilerOptions": {
         "skipLibCheck": true,
+        "noImplicitAny": true,
```
Comment (benjaminkvm) [inline languageserver/client/tsconfig.json:4]: 👍

File: languageserver/client/src/language-server.ts [hunk 4]
```diff
@@ -8,78 +10,135 @@ declare var Go
...
+  // The function that the language server can call
+  // to notify the client that the server is closing
+  onServerClose(): void
 }
 
-export async function startCadenceLanguageServer(callbacks: Callbacks) {
-  const wasm = await fetch("./languageserver.wasm")
-  const go = new Go()
-  const result = await WebAssembly.instantiateStreaming(wasm, go.importObject)
+export class CadenceLanguageServer {
```
Comment (benjaminkvm) [inline languageserver/client/src/language-server.ts:60]: 💪

File: languageserver/client/src/cadence.ts (hunk 2 above)
Comment (10thfloor) [reply languageserver/client/src/cadence.ts:127]: Anything else to add?

**Comment by turbolent** [comment]
@psiemens Thanks for the review! 🙏

**Pull request merged**